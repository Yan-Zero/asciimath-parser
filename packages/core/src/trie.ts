import { SYMBOLMAP, TokenTypes } from './symbols'

interface TokenizedValue {
  value: string
  isKeyWord: boolean
  current: number
  tex: string
  type: TokenTypes
}

const NUMBERPATTERN = /[0-9]/
const STRINGPATTERN = /\S/

type GeneTokenFn = (config: {
  current: number
  value?: string
}) => TokenizedValue

const createTextToken: GeneTokenFn = (config: {
  current: number
  value?: string
}) => {
  const { value = '', current } = config
  return { value, isKeyWord: false, current, tex: value, type: TokenTypes.Text }
}

const createConstToken: GeneTokenFn = (config: {
  current: number
  value?: string
}) => {
  const { value = '', current } = config
  return { value, isKeyWord: false, current, tex: value, type: TokenTypes.Const }
}

class Trie {
  private _root: TrieNode
  private _char_to_index: Map<string, number> = new Map()
  private _n: number

  public constructor(nodes: string[]) {
    if (nodes.length === 0)
      throw new Error('Cannot create Trie since the length of nodes is 0')

    nodes.forEach((n) => {
      if (n.length !== 1)
        throw new Error(`Value \`${n}\` is invalid, the length of char must be 1`)
    })

    const dedupped = Array.from(new Set(nodes))

    this._n = dedupped.length
    this._root = new TrieNode(this._n)
    dedupped.forEach((ch, i) => {
      this._char_to_index.set(ch, i)
    })
  }

  private c2i(key: string) {
    return this._char_to_index.get(key)
  }

  public insert(key: string) {
    if (key.length === 0)
      return

    let root = this._root

    ;([...key]).forEach((ch, i) => {
      const idx = this.c2i(ch)
      if (typeof idx === 'undefined')
        throw new Error(`key \`${ch}\` not in key set`)
      if (root._nextNode[idx] === null)
        root._nextNode[idx] = new TrieNode(this._n)
      root = root._nextNode[idx]!
      if (i === key.length - 1)
        root._end = true
    })
  }

  public search(word: string): boolean {
    if ((!this._root._nextNode.find(i => i !== null)) || word.length === 0)
      return false
    let root = this._root
    let i = 0
    for (; i < word.length; i++) {
      const ch = word[i]
      const idx = this.c2i(ch)
      if (typeof idx === 'undefined')
        throw new Error(`key \`${ch}\` not in key set`)
      if (root._nextNode[idx] === null)
        return false
      root = root._nextNode[idx]!
    }
    if (i === word.length)
      return true
    return false
  }

  /**
   * @param letters input string
   * @returns value: matched keyword
   * current: the string cursor
   */
  public tryParsing(letters: string[], start = 0): TokenizedValue {
    let value = ''
    let root = this._root
    let isKeyWord = false
    let current = start
    // let depth = 0
    for (; current < letters.length; current++) {
      const ch = letters[current]
      // console.log('depth', depth, ch)

      // the spaces are processed in `tryParsingAll`
      // if (/\s/.test(ch)) {
      //   current++
      //   break
      // }

      const idx = this.c2i(ch)
      // cannot find key in symbol map
      if (typeof idx === 'undefined')
        break

      if (root._nextNode[idx] === null)
        break

      value += ch
      root = root._nextNode[idx]!
      isKeyWord = root._end
      // depth++
    }
    const ret = (() => {
      if (isKeyWord)
        return SYMBOLMAP.get(value)!
      return { tex: value, type: TokenTypes.StringLiteral }
    })()
    return { value, isKeyWord, current, ...ret }
  }

  public tryParsingNumber(letters: string[], current: number): TokenizedValue {
    let ch = letters[current]
    let value = ''
    while (NUMBERPATTERN.test(ch) && current < letters.length) {
      value += ch
      ch = letters[++current]
    }
    if (ch === '.') {
      value += ch
      ch = letters[++current]
    }
    while (NUMBERPATTERN.test(ch) && current < letters.length) {
      value += ch
      ch = letters[++current]
    }
    return { value, isKeyWord: false, current, tex: value, type: TokenTypes.NumberLiteral }
  }

  public tryParsingString(letters: string[], current: number): TokenizedValue {
    let ch = letters[current]
    let value = ''
    while (STRINGPATTERN.test(ch) && current < letters.length) {
      const idx = this.c2i(ch)
      if (typeof idx !== 'undefined' && this._root._nextNode[idx] !== null)
        break
      value += ch
      ch = letters[++current]
    }
    return { value, isKeyWord: false, current, tex: value, type: TokenTypes.StringLiteral }
  }

  public tryParsingNewLines(letters: string[], current: number): TokenizedValue {
    let ch = letters[current]
    let value = ''
    while (/\n/.test(ch) && current < letters.length) {
      value += ch
      ch = letters[++current]
    }
    if (value.length >= 2)
      return { value, isKeyWord: true, current, tex: '\\\\', type: TokenTypes.Align }
    else
      return { value: '', isKeyWord: false, current, tex: '', type: TokenTypes.None }
  }

  private getPlainTextInDoubleQuote(letters: string[], current: number): { value: string; current: number } {
    let value = ''
    let ch = letters[current]
    if (ch === '"') {
      ch = letters[++current]
      while (ch !== '"' && current < letters.length) {
        value += ch
        ch = letters[++current]
      }
      if (ch === '"') {
        current++
        return { current, value }
      }
    }
    return { value, current }
  }

  public tryParsingText(letters: string[], current: number): TokenizedValue {
    const { value, current: newCurrent } = this.getPlainTextInDoubleQuote(letters, current)
    return { value, isKeyWord: false, current: newCurrent, tex: value, type: TokenTypes.Text }
  }

  private getColorString(letters: string[], current: number): TokenizedValue {
    let color = ''
    while (current < letters.length) {
      const ch = letters[current]
      if (!/[#\da-z]/i.test(ch))
        break
      color += ch
      current++
    }

    return { value: color, isKeyWord: false, current, tex: color, type: TokenTypes.Const }
  }

  private skipSpaces(letters: string[], current: number): number {
    while (current < letters.length) {
      const ch = letters[current]
      if (!/\s/.test(ch))
        break
      current++
    }
    return current
  }

  private processColor(letters: string[], current: number): TokenizedValue {
    let existParen = false
    let res = { value: '', isKeyWord: false, current, tex: '', type: TokenTypes.Const }
    current = this.skipSpaces(letters, current)
    if (current >= letters.length)
      return res
    { // detect paren
      const ch = letters[current]
      if (/[\(\{\[]/.test(ch)) {
        existParen = true
        current++
      }
    }
    current = this.skipSpaces(letters, current)
    res = this.getColorString(letters, current)
    current = res.current
    current = this.skipSpaces(letters, current)
    res.current = current
    if (current >= letters.length)
      return res
    {
      const ch = letters[current]
      if (/[\)\}\]]/.test(ch) && existParen)
        current++
    }
    res.current = current
    return res
  }

  private getPlainText(letters: string[], current: number, fn: GeneTokenFn): TokenizedValue {
    let useParen = false
    let ch = letters[current]
    while (/\s/.test(ch))
      ch = letters[++current]

    useParen = ch === '('
    let value = ''
    if (useParen) {
      ch = letters[++current]
      while (current < letters.length && ch !== ')') {
        value += ch
        ch = letters[++current]
      }
      current++
      return fn({ current, value })
    }
    // do not use paren, then read a word
    while (current < letters.length && /\S/.test(ch)) {
      value += ch
      ch = letters[++current]
    }
    return fn({ current, value })
  }

  public tryParsingAll(word: string) {
    let current = 0
    const tokens: TokenizedValue[] = []
    let counter = 0
    const letters = [...word]
    while (current < letters.length) {
      {
        const t = this.tryParsingNewLines(letters, current)
        current = t.current
        if (t.value !== '') {
          tokens.push(t)
          continue
        }
      }
      // process spaces
      if (/\s/.test(letters[current])) {
        current++
        continue
      }
      // process potential keywords
      let t = this.tryParsing(letters, current)
      current = t.current
      if (t.value !== '') {
        switch (t.value) {
          case 'text': {
            t = this.getPlainText(letters, current, createTextToken)
            current = t.current
            tokens.push(t)
            break
          }
          case 'tex': {
            // t = this.getPlainText(letters, current, createConstToken)
            current = this.skipSpaces(letters, current)
            const { value: v, current: c } = this.getPlainTextInDoubleQuote(letters, current)
            current = c
            t = createConstToken({ current, value: v })
            tokens.push(t)
            break
          }
          case 'hspace': {
            tokens.push(t)
            const nt = this.getPlainText(letters, current, createConstToken)
            current = nt.current
            tokens.push(nt)
            break
          }
          case 'color': {
            tokens.push(t)
            const nt = this.processColor(letters, current)
            current = nt.current
            tokens.push(nt)
            break
          }
          default: {
            tokens.push(t)
          }
        }
        continue
      }
      // process numbers
      {
        const t = this.tryParsingNumber(letters, current)
        current = t.current
        if (t.value !== '') {
          tokens.push(t)
          continue
        }
      }
      // process text wrapped with double quote
      {
        const t = this.tryParsingText(letters, current)
        current = t.current
        if (t.value !== '') {
          tokens.push(t)
          continue
        }
      }
      // process other string
      {
        const t = this.tryParsingString(letters, current)
        current = t.current
        if (t.value !== '') {
          tokens.push(t)
          continue
        }
      }
      counter++
      if (counter > letters.length * 2)
        throw new Error('Oops! There may be an infinity loop')
    }
    return tokens
  }
}

class TrieNode {
  _nextNode: (TrieNode | null)[] = []
  _end = false
  constructor(n: number) {
    this._nextNode = Array.from({ length: n }, () => null)
  }
}

function createTrie(config: {
  extConst?: Array<[string, string]>
} = {}) {
  const charset: Set<string> = new Set([])
  config.extConst?.forEach(([k, v]) => {
    SYMBOLMAP.set(k, { type: TokenTypes.Const, tex: v })
  })
  for (const k of SYMBOLMAP.keys())
    [...k].forEach(i => charset.add(i))
  const chars = Array.from(charset)
  chars.push(' ')

  const trie = new Trie(chars)
  for (const k of SYMBOLMAP.keys())
    trie.insert(k)

  return trie
}

export {
  Trie,
  TrieNode,
  createTrie,
  TokenizedValue,
}
