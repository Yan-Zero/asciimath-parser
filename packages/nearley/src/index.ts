import Nearley from 'nearley'
import initGrammar from './grammar.js'
import initLexer from './lexer.js'
import type { Symbols } from './symbols.js'
import initSymbols from './symbols.js'
import initGenerator from './to-tex.js'

export type Ast = any
type ReplaceLaw = [RegExp | string, string | ((substring: string, ...args: any[]) => string)]

interface AsciiMathConfig {
  /**
   * @default true
   * enable displayMode in KaTeX
   */
  display?: boolean
  /**
   * @default false
   * throws error on parse
   */
  throws?: boolean
  /**
   * Add custom symbols.
   *
   * For example:
   * {
   *   keyword: {
   *     dx: { tex: '{\\text{d}x}' },
   *     dy: { tex: '{\\text{d}y}' },
   *     dz: { tex: '{\\text{d}z}' },
   *     dt: { tex: '{\\text{d}t}' },
   *     ee: { tex: '\\text{e}' },
   *     ii: { tex: '\\text{i}' },
   *   }
   * }
   */
  symbols?: Symbols
  /**
   * Replace target expressions before tokenizing
   *
   * For example:
   * [
   *   [/&#(x?[0-9a-fA-F]+);/g, (match, $1) =>
   *     String.fromCodePoint($1[0] === 'x' ? '0' + $1 : $1)
   *   ],
   *   ...
   * ]
   */
  replaceBeforeTokenizing?: ReplaceLaw[]
}

type RestrictedAmConfig = Required<AsciiMathConfig>

function resolveConfig(config?: AsciiMathConfig): RestrictedAmConfig {
  return {
    display: config?.display ?? true,
    throws: config?.throws ?? false,
    symbols: {
      keyword: {
        dx: { tex: '{\\text{d}x}' },
        dy: { tex: '{\\text{d}y}' },
        dz: { tex: '{\\text{d}z}' },
        dt: { tex: '{\\text{d}t}' },
        ee: { tex: '\\text{e}' },
        ii: { tex: '\\text{i}' },
      },
      ...config?.symbols,
    },
    replaceBeforeTokenizing: [
      // html entity
      [/&#(x?[0-9a-fA-F]+);/g, (_match, $1) =>
        String.fromCodePoint($1[0] === 'x' ? `0${$1}` : $1),
      ],
      ...(config?.replaceBeforeTokenizing || []),
    ],
  }
}

class AsciiMath {
  public display: boolean
  public throws: boolean
  public replaceBeforeTokenizing: ReplaceLaw[]
  public lexer: Nearley.Lexer
  public parser: Nearley.Parser
  private genTex: ((ast: Ast) => string)
  private initState: { [key: string]: any; lexerState: Nearley.LexerState }
  constructor(config?: AsciiMathConfig) {
    const { display, throws, symbols, replaceBeforeTokenizing } = resolveConfig(config)
    this.display = display
    this.throws = throws
    this.replaceBeforeTokenizing = replaceBeforeTokenizing
    const allSymbols = initSymbols(symbols)
    const lexer = this.lexer = initLexer(allSymbols)
    const grammar = initGrammar(lexer)
    const compiledGrammar = Nearley.Grammar.fromCompiled(grammar)
    this.parser = new Nearley.Parser(compiledGrammar)
    this.initState = this.parser.save()
    this.genTex = initGenerator(allSymbols)
  }

  toTex(code: string): string {
    this.parser.restore(this.initState)
    try {
      code = this.replaceBeforeTokenizing.reduce((prev, curLaw) => {
        // nodejs 14 没有 replaceAll 方法, 不要用 replaceAll
        return prev.replace(new RegExp(curLaw[0], 'g'), curLaw[1] as any)
      }, code)
      this.parser.feed(code)
      if (this.parser.results.length === 0) {
        throw new Error('unexpected end of input')
      }
      else if (this.parser.results.length > 1) {
        console.error(this.parser.results)
        throw new Error('ambiguous parse')
      }
      let res = this.genTex(this.parser.results)
      if (this.display)
        res = `\\displaystyle{ ${res} }`
      return res
    }
    catch (e) {
      if (this.throws)
        throw e
      console.error(e)
      const message = String(e)
      let index = message.indexOf(' Instead, I was expecting to see one of the following:')
      if (index > -1) {
        return this.genTex({
          type: 'opOA',
          value: 'verb',
          $1: { value: message.slice(0, index) },
        })
      }

      index = message.indexOf(':\n')
      const end = index > -1 ? index : undefined
      return `\\text{${message.slice(0, end)}}`
    }
  }
}

export {
  AsciiMath,
  AsciiMathConfig,
}
