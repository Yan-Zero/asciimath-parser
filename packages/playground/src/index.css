@import url('https://fonts.googleapis.com/css2?family=JetBrains+Mono&display=swap');

:root {
  font-family: Inter, Avenir, Helvetica, Arial, sans-serif;
  font-size: 16px;
  line-height: 24px;
  font-weight: 400;

  color-scheme: light dark;
  color: rgba(255, 255, 255, 0.87);
  background-color: #242424;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}

a {
  font-weight: 500;
  color: #646cff;
  text-decoration: inherit;
}
a:hover {
  color: #535bf2;
}

body {
  margin: 0;
  min-width: 320px;
  min-height: 100vh;
}

h1 {
  font-size: 3.2em;
  line-height: 1.1;
  margin-block-end: 0.25rem;
}

h2 {
  margin-top: 3rem;
}

h3 {
  margin-top: 2rem;
}

button {
  border-radius: 8px;
  border: 1px solid transparent;
  padding: 0.6em 1.2em;
  font-size: 1em;
  font-weight: 500;
  font-family: inherit;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.25s;
  display: flex;
  align-items: center;
  justify-content: center;
}
button:hover {
  border-color: #646cff;
}
button:focus,
button:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

@media (prefers-color-scheme: light) {
  :root {
    color: #213547;
    background-color: #ffffff;
  }
  a:hover {
    color: #747bff;
  }
  button {
    background-color: #f9f9f9;
  }
}

td {
  padding: 0.5rem;
  white-space: nowrap;
}

table {
  width: 100%;
  border-top: 2px solid;
  border-bottom: 2px solid;
}

.tc {
  text-align: center;
}

.tex {
  background-color: #7f7f7f33;
}

main {
  max-width: 800px;
  margin: 0 auto;
  padding: 1rem;
}

.space-x-1 > :not([hidden]) ~ :not([hidden]) {
  --tw-space-x-reverse: 0;
  margin-right: calc(0.25rem * var(--tw-space-x-reverse));
  margin-left: calc(0.25rem * calc(1 - var(--tw-space-x-reverse)));
}

.flex {
  display: flex;
}
.w-full {
  width: 100%;
}
.mb-4 {
  margin-bottom: 1rem;
}

code, pre {
  font-family: 'JetBrains Mono', 'Courier New', Courier, monospace;
  font-size: 0.8rem;
}

code {
  background-color: #7f7f7f11;
  padding: 0.2rem;
  border-radius: 0.2rem;
}

pre {
  background-color: #7f7f7f11;
  padding: 0.75em;
  border-radius: 0.2rem;
  overflow: auto;
}