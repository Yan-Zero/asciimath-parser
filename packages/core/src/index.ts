import { codegen } from './codegen'
import { parser } from './parser'
import type { Trie } from './trie'
import { createTrie } from './trie'

type ReplaceLaw = [RegExp | string, string | ((substring: string, ...args: any[]) => string)]

interface AsciiMathConfig {
  /**
   * @default true
   * enable displayMode in KaTeX
   */
  display?: boolean
  /**
   * Translate custom keywords into LaTeX expressions
   *
   * For example:
   * [
   *   ['dx', '\text{d}x'],
   *   ['dy', '\text{d}y']
   * ]
   */
  extConst?: Array<[string, string]>
  /**
   * Replace target expressions before tokenizing
   *
   * For example:
   * [
   *   [/&#(x?[0-9a-fA-F]+);/g, (match, $1) =>
   *     String.fromCodePoint($1[0] === 'x' ? '0' + $1 : $1)
   *   ],
   *   ...
   * ]
   */
  replaceBeforeTokenizing?: ReplaceLaw[]
}

type RestrictedAmConfig = Required<AsciiMathConfig>

function resolveConfig(config?: AsciiMathConfig): RestrictedAmConfig {
  const defaultConfig: RestrictedAmConfig = {
    display: true,
    extConst: [
      ['dx', '{\\text{d}x}'],
      ['dy', '{\\text{d}y}'],
      ['dz', '{\\text{d}z}'],
      ['dt', '{\\text{d}t}'],
      ['#', '\\displaystyle'],
    ],
    replaceBeforeTokenizing: [
      [/&#(x?[0-9a-fA-F]+);/g, (_match, $1) =>
        String.fromCodePoint($1[0] === 'x' ? `0${$1}` : $1),
      ],
    ],
  }
  if (typeof config?.display !== 'undefined')
    defaultConfig.display = config?.display
  if (config?.extConst?.length)
    defaultConfig.extConst.push(...config.extConst)
  if (config?.replaceBeforeTokenizing?.length)
    defaultConfig.replaceBeforeTokenizing.push(...config.replaceBeforeTokenizing)

  return defaultConfig
}

class AsciiMath {
  private trie: Trie
  private display: boolean
  private replaceLaws: ReplaceLaw[]
  constructor(config?: AsciiMathConfig) {
    const { display, extConst, replaceBeforeTokenizing: replaceBeforeParsing } = resolveConfig(config)
    this.trie = createTrie({ extConst })
    this.display = display
    this.replaceLaws = replaceBeforeParsing
  }

  toTex(code: string): string {
    try {
      code = this.replaceLaws.reduce((prev, curLaw) => {
      // in fact the judgement is no use...
        if (typeof curLaw[1] === 'function')
          return prev.replace(curLaw[0], curLaw[1])
        else
          return prev.replace(curLaw[0], curLaw[1])
      }, code)
      let res = codegen(parser(this.trie.tryParsingAll(code)))
      if (this.display)
        res = `\\displaystyle{ ${res} }`
      return res
    }
    catch (e) {
      return `\\text{${String(e)}}`
    }
  }
}

export {
  AsciiMath,
  AsciiMathConfig,
}
