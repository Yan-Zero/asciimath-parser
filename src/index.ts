import { codegen } from './codegen'
import { parser } from './parser'
import type { Trie } from './trie'
import { createTrie } from './trie'

type ReplaceLaw = [RegExp | string, string | ((substring: string, ...args: any[]) => string)]

interface AsciiMathConfig {
  /**
   * @default true
   * enable displayMode in KaTeX
   */
  display?: boolean
  /**
   * Translate custom keywords into LaTeX expressions
   *
   * For example:
   * [
   *   ['dx', '\text{d}x'],
   *   ['dy', '\text{d}y']
   * ]
   */
  extConst?: Array<[string, string]>
  /**
   * Replace target expressions before tokenizing
   *
   * For example:
   * [
   *   [/&#(x?[0-9a-fA-F]+);/g, (match, $1) =>
   *     String.fromCodePoint($1[0] === 'x' ? '0' + $1 : $1)
   *   ],
   *   ...
   * ]
   */
  replaceBeforeTokenizing?: ReplaceLaw[]
}

type RestrictedAmConfig = Required<AsciiMathConfig>

function resolveConfig(config?: AsciiMathConfig): RestrictedAmConfig {
  const defaultConfig: RestrictedAmConfig = {
    display: true,
    extConst: [
      ['dx', '{\\text{d}x}'],
      ['dy', '{\\text{d}y}'],
      ['dz', '{\\text{d}z}'],
      ['dt', '{\\text{d}t}'],
    ],
    replaceBeforeTokenizing: [
      [/part(\^\S*)?\s+(\S+)\s+(\([^)]*\)|\S+)/g, (_match, $1, $2, $3) => {
        if (!$1)
          $1 = ''
        if ($3[0] === '(')
          $3 = $3.slice(1, -1).split(/\s+/).join(' del ')
        return `(del${$1} ${$2})/(del ${$3})`
      }],
      [/&#(x?[0-9a-fA-F]+);/g, (_match, $1) =>
        String.fromCodePoint($1[0] === 'x' ? `0${$1}` : $1),
      ],
    ],
  }
  if (config?.display)
    defaultConfig.display = config?.display
  if (config?.extConst?.length)
    defaultConfig.extConst.push(...config.extConst)
  if (config?.replaceBeforeTokenizing?.length)
    defaultConfig.replaceBeforeTokenizing.push(...config.replaceBeforeTokenizing)

  return defaultConfig
}

class AsciiMath {
  private trie: Trie
  private display: boolean
  private replaceLaws: ReplaceLaw[]
  constructor(config?: AsciiMathConfig) {
    const { display, extConst, replaceBeforeTokenizing: replaceBeforeParsing } = resolveConfig(config)
    this.trie = createTrie({ extConst })
    this.display = display
    this.replaceLaws = replaceBeforeParsing
  }

  toTex(code: string): string {
    try {
      code = this.replaceLaws.reduce((prev, curLaw) => {
      // in fact the judgement is no use...
        if (typeof curLaw[1] === 'function')
          return prev.replace(curLaw[0], curLaw[1])
        else
          return prev.replace(curLaw[0], curLaw[1])
      }, code)
      let res = codegen(parser(this.trie.tryParsingAll(code)))
      if (this.display)
        res = `\\displaystyle{ ${res} }`
      return res
    }
    catch (e) {
      console.error(e)
      return ''
    }
  }
}

export {
  AsciiMath,
  AsciiMathConfig,
}
