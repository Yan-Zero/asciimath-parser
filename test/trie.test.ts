import { describe, expect, it } from 'vitest'
import { SYMBOLMAP } from '../src/symbols'
import { Trie } from '../src/trie'

const MULTILINE_AM = `sum_(n=1)^(+oo)&=(pi^2/6)

f(x)&=x^2`

describe('trie', () => {
  it.skip('should build trie', () => {
    const keys = Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 97))
    keys.push(' ')
    const trie = new Trie(keys)
    trie.insert('banana')
    trie.insert('ban')
    trie.insert('abc')
    trie.insert('abda')

    expect(trie).toMatchSnapshot()
    expect(trie.search('ban')).toBe(true)
    expect(trie.search('aaaa')).toBe(false)
    // 直接搜索到关键词
    expect(trie.tryParsing('banana')).toMatchInlineSnapshot(`
      {
        "current": 6,
        "isKeyWord": true,
        "value": "banana",
      }
    `)
    // 当前单词比最长关键词长，那么到最长的关键词就结束
    expect(trie.tryParsing('bananan')).toMatchInlineSnapshot(`
      {
        "current": 6,
        "isKeyWord": true,
        "value": "banana",
      }
    `)
    expect(trie.tryParsing('bab')).toMatchInlineSnapshot(`
      {
        "current": 2,
        "isKeyWord": false,
        "value": "ba",
      }
    `)
    // 遇到空格跳过
    expect(trie.tryParsing('ab cfdgerafw')).toMatchInlineSnapshot(`
      {
        "current": 2,
        "isKeyWord": false,
        "value": "ab",
      }
    `)
    // 与关键字有相同的前缀
    expect(trie.tryParsing('bana')).toMatchInlineSnapshot(`
      {
        "current": 4,
        "isKeyWord": false,
        "value": "bana",
      }
    `)
  })

  it('should build the symbol trie', () => {
    const charset: Set<string> = new Set([])
    for (const k of SYMBOLMAP.keys())
      k.split('').forEach(i => charset.add(i))
    const chars = Array.from(charset)
    chars.push(' ')
    expect(chars).toMatchSnapshot()

    const trie = new Trie(chars)
    for (const k of SYMBOLMAP.keys())
      trie.insert(k)

    expect(trie.tryParsingAll(MULTILINE_AM)).toMatchSnapshot()
    expect(trie.tryParsingAll('x |-> "e"^(2pi "i" x)')).toMatchSnapshot()
    expect(trie.tryParsingAll('"e"^i pi')).toMatchInlineSnapshot(`
      [
        {
          "current": 3,
          "isKeyWord": false,
          "tex": "e",
          "type": "Text",
          "value": "e",
        },
        {
          "current": 4,
          "isKeyWord": true,
          "tex": "^{ $1 }",
          "type": "OperatorA",
          "value": "^",
        },
        {
          "current": 5,
          "isKeyWord": false,
          "tex": "i",
          "type": "StringLiteral",
          "value": "i",
        },
        {
          "current": 8,
          "isKeyWord": true,
          "tex": "\\\\pi",
          "type": "Const",
          "value": "pi",
        },
      ]
    `)
  })
})
