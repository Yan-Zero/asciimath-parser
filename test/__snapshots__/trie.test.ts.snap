// Vitest Snapshot v1

exports[`trie > should build the symbol trie 1`] = `
[
  "a",
  "l",
  "p",
  "h",
  "b",
  "e",
  "t",
  "g",
  "m",
  "G",
  "d",
  "D",
  "s",
  "i",
  "o",
  "n",
  "v",
  "r",
  "z",
  "T",
  "k",
  "L",
  "u",
  "x",
  "X",
  "P",
  "S",
  "c",
  "O",
  "*",
  "/",
  "\\\\",
  "|",
  ">",
  "<",
  "-",
  ":",
  "@",
  "+",
  ".",
  "^",
  "!",
  "=",
  "~",
  "f",
  "w",
  "A",
  "E",
  "_",
  "'",
  "q",
  "C",
  "N",
  "Q",
  "R",
  "Z",
  "(",
  ")",
  "[",
  "]",
  "{",
  "}",
  "&",
  ",",
  ";",
  " ",
]
`;

exports[`trie > should build the symbol trie 2`] = `
[
  {
    "current": 3,
    "isKeyWord": true,
    "tex": "\\\\sum",
    "type": "Const",
    "value": "sum",
  },
  {
    "current": 4,
    "isKeyWord": true,
    "tex": "_{ $1 }",
    "type": "OperatorA",
    "value": "_",
  },
  {
    "current": 5,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 6,
    "isKeyWord": false,
    "tex": "n",
    "type": "StringLiteral",
    "value": "n",
  },
  {
    "current": 7,
    "isKeyWord": false,
    "tex": "=",
    "type": "StringLiteral",
    "value": "=",
  },
  {
    "current": 8,
    "isKeyWord": false,
    "tex": "1",
    "type": "NumberLiteral",
    "value": "1",
  },
  {
    "current": 9,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 10,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorA",
    "value": "^",
  },
  {
    "current": 11,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 12,
    "isKeyWord": false,
    "tex": "+",
    "type": "StringLiteral",
    "value": "+",
  },
  {
    "current": 14,
    "isKeyWord": true,
    "tex": "\\\\infty",
    "type": "Const",
    "value": "oo",
  },
  {
    "current": 15,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 16,
    "isKeyWord": true,
    "tex": "&",
    "type": "Align",
    "value": "&",
  },
  {
    "current": 17,
    "isKeyWord": false,
    "tex": "=",
    "type": "StringLiteral",
    "value": "=",
  },
  {
    "current": 18,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 20,
    "isKeyWord": true,
    "tex": "\\\\pi",
    "type": "Const",
    "value": "pi",
  },
  {
    "current": 21,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorA",
    "value": "^",
  },
  {
    "current": 22,
    "isKeyWord": false,
    "tex": "2",
    "type": "NumberLiteral",
    "value": "2",
  },
  {
    "current": 23,
    "isKeyWord": true,
    "tex": "\\\\frac{ $1 }{ $2 }",
    "type": "OperatorAOB",
    "value": "/",
  },
  {
    "current": 24,
    "isKeyWord": false,
    "tex": "6",
    "type": "NumberLiteral",
    "value": "6",
  },
  {
    "current": 25,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 27,
    "isKeyWord": true,
    "tex": "\\\\\\\\",
    "type": "Align",
    "value": "

",
  },
  {
    "current": 28,
    "isKeyWord": false,
    "tex": "f",
    "type": "StringLiteral",
    "value": "f",
  },
  {
    "current": 29,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 30,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 31,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
  {
    "current": 32,
    "isKeyWord": true,
    "tex": "&",
    "type": "Align",
    "value": "&",
  },
  {
    "current": 33,
    "isKeyWord": false,
    "tex": "=",
    "type": "StringLiteral",
    "value": "=",
  },
  {
    "current": 34,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 35,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorA",
    "value": "^",
  },
  {
    "current": 36,
    "isKeyWord": false,
    "tex": "2",
    "type": "NumberLiteral",
    "value": "2",
  },
]
`;

exports[`trie > should build the symbol trie 3`] = `
[
  {
    "current": 1,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 5,
    "isKeyWord": true,
    "tex": "\\\\mapsto",
    "type": "Const",
    "value": "|->",
  },
  {
    "current": 9,
    "isKeyWord": false,
    "tex": "e",
    "type": "Text",
    "value": "e",
  },
  {
    "current": 10,
    "isKeyWord": true,
    "tex": "^{ $1 }",
    "type": "OperatorA",
    "value": "^",
  },
  {
    "current": 11,
    "isKeyWord": true,
    "tex": "(",
    "type": "LParen",
    "value": "(",
  },
  {
    "current": 12,
    "isKeyWord": false,
    "tex": "2",
    "type": "NumberLiteral",
    "value": "2",
  },
  {
    "current": 14,
    "isKeyWord": true,
    "tex": "\\\\pi",
    "type": "Const",
    "value": "pi",
  },
  {
    "current": 18,
    "isKeyWord": false,
    "tex": "i",
    "type": "Text",
    "value": "i",
  },
  {
    "current": 20,
    "isKeyWord": false,
    "tex": "x",
    "type": "StringLiteral",
    "value": "x",
  },
  {
    "current": 21,
    "isKeyWord": true,
    "tex": ")",
    "type": "RParen",
    "value": ")",
  },
]
`;
